packages/frontend/src/main/components/common/PreviewImage.vue

```diff
@@ -27,33 +35,119 @@ export default {
+        22, 23, 0
+      ]
     }
   },
-  mounted() {
-    this.getPreviewImage().then().catch()
+  computed: {
+    revImg() {
+      return this.legacyMode ? this.previewImages : [...this.previewImages].reverse()
+    },
+    bgStyle() {
+      let bgStyle = `
+        height: 100%;
+        width: 100%;
+        background-size: cover;
+        background-image:`
+
+      for (let i = 0; i < this.revImg.length; i++) {
+        bgStyle += `url(${this.revImg[i]})${i !== this.revImg.length - 1 ? ',' : ';'}`
+      }
+      bgStyle += `
+      background-position:`
+      for (let i = 0; i < this.revImg.length; i++) {
+        bgStyle += `${i === this.imageIndex ? 'center' : '10000px'}${
+          i !== this.revImg.length - 1 ? ',' : ';'
+        }`
+      }
+      return bgStyle
+    }
+  },
+
+  watch: {
+    hovered(val) {
+      if (!this.rotate) return
+      if (!val || this.hasStartedLoadingImages) return
+      this.getOtherAngles()
+      this.hasStartedLoadingImages = true
+    }
+  },
+  async mounted() {
+    this.previewImages.push(await this.getPreviewImage())
   },
   methods: {
-    async getPreviewImage() {
-      const res = await fetch(this.url, {
+    parseTouch(e) {
+      if (this.hasStartedLoadingImages && this.loading && !this.rotate) return
+      this.hovered = true
+      this.setIndex({
+        target: e.target,
+        clientX: e.touches[0].clientX,
+        clientY: e.touches[0].clientY
+      })
+    },
+    setIndex(e) {
+      if (this.hasStartedLoadingImages && this.loading && !this.rotate) return
+      const rect = e.target.getBoundingClientRect()
+      const x = e.clientX - rect.left
+      const step = rect.width / this.previewImages.length
+      let index = Math.round(x / step)
+      if (index >= this.previewImages.length) index = 0
+      this.imageIndex = index
+    },
+    async getPreviewImage(angle = 0) {
+      const res = await fetch(this.url + `/${angle}`, {
         headers: localStorage.getItem('AuthToken')
           ? { Authorization: `Bearer ${localStorage.getItem('AuthToken')}` }
           : {}
       })
-      try {
-        const blob = await res.blob()
-        this.currentPreviewImg = URL.createObjectURL(blob)
-      } catch (err) {
-        // console.log(err)
+
+      if (res.headers.has('X-Preview-Error')) {
+        throw new Error('Failed getting preview')
+      }
+      const blob = await res.blob()
+      return URL.createObjectURL(blob)
+    },
+    async getOtherAngles() {
+      // Note: previously, previews were generated for only 5 angles (-30deg, -15, 0, 15, 30deg), corresponding to
+      // labelled angles (-2, -1, 0, 1, 2). We have now switched to generating full 360deg previews in increments
+      // of 15 deg, going clockwise straight. Eg., 0 = 0deg, 1 = 15deg, 2 = 30 deg, etc.
+      this.loading = true
+      // starting in reverse to quickly figure out if we're on the legacy track!
+      for (let i = this.angles.length - 1; i > 0; i--) {
```

Comment:
It'll def be faster to load all of the images concurrently, instead of awaiting each `fetch` call sequentially.

e.g.
```js
const images = await Promise.all([
  this.getPreviewImage(0),
this.getPreviewImage(1),
this.getPreviewImage(2),
...
]);
```

Reply:
ah I see, yea you can do it that way. since the amount of angles (images) is pretty big, even if you do the first 2 sequentially, making all of the other ones concurrent should help a lot

Reply:
btw it's worth testing out how this will act on a slow network (in dev tools you can usually throttle your network connection to 3G, for example), another concern I have is that `loading` is set to `false` only after all images are loaded and there are quite a lot. might be more performant to show the images you have, while the other ones are loading

Reply:
yea, i tested on slow 3g - not going work magic there. we display the first one by default anyway; the rest of the angles start loading after a hover/touchstart event.